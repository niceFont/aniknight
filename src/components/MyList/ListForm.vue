<template>
  <v-container>
    <v-row justify="center">
      <v-col cols="8">
        <v-alert elevation="1" type="error" border="left" v-if="duplicateError">Entry already exists</v-alert>
      </v-col>
    </v-row>
    <v-row class="mt-10" justify="center">
      <v-col cols="8">
        <v-form>
          <v-autocomplete
            id="t1"
            :items="items"
            :loading="loading"
            :search-input.sync="search"
            v-model="anime.title"
            outlined
            clearable
            @focus="resetError"
            hint="Enter the Anime you want to add."
            label="Title"
          ></v-autocomplete>
          <v-text-field
            id="t2"
            v-model="anime.url"
            outlined
            hint="Enter a Url to quickly access your new anime."
            label="Url"
          ></v-text-field>
          <v-btn id="t3" @click="submitAnime" dark color="indigo" depressed block>Submit</v-btn>
        </v-form>
      </v-col>
    </v-row>
  </v-container>
</template>

<script>
import { mapActions, mapState } from "vuex";

export default {
  name: "ListForm",
  data: () => ({
    anime: {
      title: "",
      url: "",
      id: 0
    },
    loading: false,
    suggestions: [],
    search: null,
    duplicateError: null
  }),
  computed: {
    ...mapState({
      myList: state => state.myList
    }),
    items() {
      return this.suggestions.map(suggestion => {
        return `${suggestion.title.userPreferred} (${suggestion.startDate.year})`;
      });
    },
    selected() {
      return this.suggestions.filter(suggestion => {
        const selection =
          suggestion.title.userPreferred ===
          this.anime.title.slice(0, this.anime.title.length - 6).trim();

        return selection;
      })[0];
    }
  },
  methods: {
    ...mapActions(["addAnime"]),
    resetError() {
      this.duplicateError = null;
    },
    checkForDuplicate(arr) {
      return arr.every(anime => anime.id !== this.selected.id);
    },
    formatUrl(url) {
      if (url.startsWith("http://") || url.startsWith("https://")) {
        return url;
      }
      return "http://" + url;
    },
    submitAnime() {
      const notduplicate = this.checkForDuplicate(this.myList);
      if (notduplicate) {
        this.$store.dispatch("addAnime", {
          title: this.selected.title.userPreferred,
          url: this.formatUrl(this.anime.url),
          id: this.selected.id,
          src: this.selected.coverImage.medium
        });
        this.clearAll();
      } else {
        this.duplicateError = true;
      }
    },
    clearAll() {
      this.adding = false;
      this.anime.title = "";
      this.anime.url = "";
    }
  },
  watch: {
    search(value) {
      this.loading = true;
      fetch("https://graphql.anilist.co", {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Accept: "application/json"
        },
        body: JSON.stringify({
          query: `query ($page: Int = 1, $search: String) {
  Page(page: $page, perPage: 5) {
    ANIME: media(search: $search) {
      id
      title {
        userPreferred
      }
      coverImage {
        medium
      }
      startDate {
        year
      }
      type
    }
  }
}`,
          variables: { search: value }
        })
      })
        .then(res => res.json())
        .then(res => {
          if (res.data.Page) this.suggestions = [...res.data.Page.ANIME];
        })
        .catch(err => console.error(err))
        .finally(() => (this.loading = false));
    }
  }
};
</script>

<style>
#error {
  color: red;
}
</style>