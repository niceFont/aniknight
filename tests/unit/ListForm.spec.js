import { mount, shallowMount, createLocalVue } from "@vue/test-utils"
import ListForm from "../../src/components/MyList/ListForm.vue"
import Vue from "vue"
import Vuetify from "vuetify"

describe("ListForm.vue", () => {

    // let vuetify
    let wrapper

    beforeEach(() => {
        //const localVue = createLocalVue()
        Vue.use(Vuetify)
        wrapper = shallowMount(ListForm)
    })

    it("renders", () => {
        expect(shallowMount(ListForm).isVueInstance()).toBe(true)

    })

    it("has a div", () => {
        expect(wrapper.contains("div")).toBe(true)
    })
})