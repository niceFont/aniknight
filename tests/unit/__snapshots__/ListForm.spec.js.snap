// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListForm.vue Matches snapshot 1`] = `
<div
  class="container"
>
  <div
    class="row justify-center"
  >
    <div
      class="col col-8"
    >
      <!---->
    </div>
  </div>
   
  <div
    class="row mt-10 justify-center"
  >
    <div
      class="col col-8"
    >
      <form
        class="v-form"
        novalidate="novalidate"
      >
        <div
          class="v-input theme--light v-text-field v-text-field--enclosed v-text-field--outlined v-select v-autocomplete"
        >
          <div
            class="v-input__control"
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-owns="list-27"
              class="v-input__slot"
              role="combobox"
            >
              <fieldset
                aria-hidden="true"
              >
                <legend
                  style="width: 0px;"
                >
                  <span>
                    ​
                  </span>
                </legend>
              </fieldset>
              <div
                class="v-select__slot"
              >
                <label
                  class="v-label theme--light"
                  for="t1"
                  style="right: 0px; position: absolute;"
                >
                  Title
                </label>
                <input
                  autocomplete="off"
                  id="t1"
                  type="text"
                />
                <div
                  class="v-input__append-inner"
                >
                  <div
                    class="v-input__icon v-input__icon--"
                  >
                    <i
                      class="v-icon notranslate v-icon--link material-icons theme--light"
                      role="button"
                    />
                  </div>
                </div>
                <div
                  class="v-input__append-inner"
                >
                  <div
                    class="v-input__icon v-input__icon--append"
                  >
                    <i
                      aria-hidden="true"
                      class="v-icon notranslate material-icons theme--light"
                    >
                      $dropdown
                    </i>
                  </div>
                </div>
                <input
                  type="hidden"
                  value="anime (2012)"
                />
              </div>
              <div
                class="v-menu"
              >
                <div
                  class="v-menu__content theme--light v-autocomplete__content"
                  style="max-height: 304px; min-width: 0px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;"
                />
              </div>
            </div>
            <div
              class="v-text-field__details"
            >
              <div
                class="v-messages theme--light"
              >
                <span
                  class="v-messages__wrapper"
                  name="message-transition"
                  tag="div"
                />
              </div>
            </div>
          </div>
        </div>
         
        <div
          class="v-input theme--light v-text-field v-text-field--enclosed v-text-field--outlined"
        >
          <div
            class="v-input__control"
          >
            <div
              class="v-input__slot"
            >
              <fieldset
                aria-hidden="true"
              >
                <legend
                  style="width: 0px;"
                >
                  <span>
                    ​
                  </span>
                </legend>
              </fieldset>
              <div
                class="v-text-field__slot"
              >
                <label
                  class="v-label theme--light"
                  for="t2"
                  style="right: 0px; position: absolute;"
                >
                  Url
                </label>
                <input
                  id="t2"
                  type="text"
                />
              </div>
            </div>
            <div
              class="v-text-field__details"
            >
              <div
                class="v-messages theme--light"
              >
                <span
                  class="v-messages__wrapper"
                  name="message-transition"
                  tag="div"
                />
              </div>
            </div>
          </div>
        </div>
         
        <button
          class="v-btn v-btn--block v-btn--depressed theme--dark v-size--default indigo"
          id="t3"
          type="button"
        >
          <span
            class="v-btn__content"
          >
            Submit
          </span>
        </button>
      </form>
    </div>
  </div>
</div>
`;
